File: src/main/java/com/example/its/domain/auth/UserService.java
package com.example.its.domain.auth;

import lombok.RequiredArgsConstructor;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @PreAuthorize("hasAuthority('ADMIN')")
    public List<User> findAll() {
        return userRepository.findAll();
    }

    @PreAuthorize("hasAuthority('ADMIN')")
    public void create(String username, String password, String authority) {
        var encodedPassword = passwordEncoder.encode(password);
        userRepository.insert(username, encodedPassword, authority);
    }

    @Transactional
    @PreAuthorize("hasAuthority('ADMIN')")
    public void update(String targetUsername, String newUsername, String rawPassword, String authority) {
        String encodedPassword = passwordEncoder.encode(rawPassword);
        userRepository.update(targetUsername, newUsername, encodedPassword, authority);
    }

    @Transactional
    @PreAuthorize("hasAuthority('ADMIN')")
    public void delete(String username) {
        userRepository.delete(username);
    }
}

File: frontend/src/services/userService.ts
import type { User, UserForm } from '../types';
import { apiClient } from './apiClient';

export const userService = {
  getAll: async (): Promise<User[]> => {
    const response = await apiClient.get<User[]>('/users');
    return response.data;
  },

  create: async (newUser: UserForm): Promise<User> => {
    const response = await apiClient.post<User>('/users', newUser);
    return response.data;
  },

  update: async (
    targetUsername: string,
    data: { username: string; password?: string; authority: string },
  ): Promise<User> => {
    const response = await apiClient.put<User>(`/users/${targetUsername}`, {
      username: data.username,
      password: data.password,
      authority: data.authority,
    });

    return response.data;
  },

  delete: async (username: string): Promise<void> => {
    await apiClient.delete(`/users/${username}`);
  },
};
